### PR TITLE
add alt-score

### PR DESCRIPTION
**Pull request by malaffoon**
This adds the alt-score aggregate report to the backend services. 

It became apparent that the RSTs have not been run in a long while. They were failing for all sorts of reasons having nothing to do with my changes. Annoying, but i think they are cleaned up now. They are still flaky enough that i'm not sure i'll enable them in CI/CD builds.

### CODE REVIEW

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcAltScoreReportRepository.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                    .avgScaleScore(getNullable(row, row.getInt("score")))
+                    .avgStdErr(getNullable(row, row.getInt("std_err")))
+                    .level1Count(row.getInt("level1"))
+                    .level2Count(row.getInt("level2"))
+                    .level3Count(row.getInt("level3"))
+                    .level4Count(row.getInt("level4"))
+                    .level5Count(row.getInt("level5"))
+                    .level6Count(row.getInt("level6"))
+                    .build())
+                .altScoreCode(row.getString("alt_score_code"))
```
Comment (malaffoon) [inline aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcAltScoreReportRepository.java:106]: This is a core bit of code: it extracts the row from the query results.

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcClaimReportRepository.java [hunk 2]
```diff
@@ -159,7 +158,7 @@ public int estimateReportRowCountByOrganizationType(@NotNull final ClaimOrganiza
         final Map<ActiveAssessment, Set<String>> claimsByAssessment = getClaimsByAssessment(reportTemplate);
 
         return (reportTemplate.getOrganizations().size()
-                * claimsByAssessment.values().stream().flatMap(Collection::stream).collect(Collectors.toList()).size()
+                * (int) claimsByAssessment.values().stream().mapToLong(Collection::size).sum()
```
Comment (malaffoon) [inline aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcClaimReportRepository.java:161]: Unrelated warning fix, there are a few of these scattered about.

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/sqlbuilder/impl/query/AltScoreSqlQueryWithReportTemplateProvider.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    public Builder createBuilderWithCommonAttributes(final AltScoreOrganizationQuery query) {
+        return addFiltersAndDimension(createBuilderInstance(), query)
+                .reportTemplate(templateProvider.createBuilderWithCommonAttributes(query).build())
+                .parameter("school_years", padIntToSize(query.getSchoolYears(), getRepositoryHelper().maxSchoolYears()))
+                .parameter("asmt_grade_ids", padIntToSize(getRepositoryHelper().toGradeIds(query.getAssessmentGradeCodes()), getRepositoryHelper().maxAssessmentGrades()))
+                .parameter("subject_ids", getRepositoryHelper().toSubjectIds(query.getAltScoreCodesBySubject().keySet()))
+                .parameter("asmt_type_id", getRepositoryHelper().toAssessmentTypeId(query.getAssessmentTypeCode()))
+                .parameter("subject_alt_score_codes", padAltScoreCodes(query.getAltScoreCodesBySubject().values().stream().flatMap(Collection::stream).collect(toList())));
```
Comment (malaffoon) [inline aggregate-service/src/main/java/org/opentestsystem/rdw/olap/sqlbuilder/impl/query/AltScoreSqlQueryWithReportTemplateProvider.java:59]: Another core bit: pass alt score codes to the query.

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/sqlbuilder/impl/query/ClaimSqlQueryWithReportTemplateProvider.java [hunk 4]
```diff
@@ -55,19 +55,10 @@ public Builder createBuilderWithCommonAttributes(final ClaimOrganizationQuery qu
...
-        if (collection != null && collection.size() >= claimsSqlPadding.size()) return collection;
-
-        if (collection == null) return claimsSqlPadding;
-        final List<String> padded = newArrayList(claimsSqlPadding);
-
-        final Iterator<String> iterator = collection.iterator();
-        for (int i = 0; i < collection.size(); ++i) {
-            padded.set(i, iterator.next());
-        }
-        return padded;
```
Comment (malaffoon) [inline aggregate-service/src/main/java/org/opentestsystem/rdw/olap/sqlbuilder/impl/query/ClaimSqlQueryWithReportTemplateProvider.java:71]: There was an existing helper that did this, use it instead of duplicating code.

File: aggregate-service/src/main/resources/alt-score-report-template.sql.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            sum(CASE WHEN eas.performance_level = 1 THEN 1 ELSE 0 END) AS level1,
+            sum(CASE WHEN eas.performance_level = 2 THEN 1 ELSE 0 END) AS level2,
+            sum(CASE WHEN eas.performance_level = 3 THEN 1 ELSE 0 END) AS level3,
+            sum(CASE WHEN eas.performance_level = 4 THEN 1 ELSE 0 END) AS level4,
+            sum(CASE WHEN eas.performance_level = 5 THEN 1 ELSE 0 END) AS level5,
+            sum(CASE WHEN eas.performance_level = 6 THEN 1 ELSE 0 END) AS level6,
+            e.school_year,
+            e.asmt_id,
+            eas.subject_score_id,
+            ss.code as alt_score_code
```
Comment (malaffoon) [inline aggregate-service/src/main/resources/alt-score-report-template.sql.yml:19]: Another core bit: the actual query for getting the rows.

File: aggregate-service/src/main/resources/claim-report-template.sql.yml [hunk 6]
```diff
@@ -1,329 +1,36 @@
 sqlbuilder:
```
Comment (malaffoon) [inline aggregate-service/src/main/resources/claim-report-template.sql.yml:1]: Why the other query files changed so much:
1. I fixed the whitespace, the indenting was inconsistent.
2. I factored out the common bits, the addons for dimensions and filters.

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcAltScoreReportRepositoryRST.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+import static org.opentestsystem.rdw.common.model.AssessmentType.ICA;
+import static org.opentestsystem.rdw.olap.model.AggregateServicePermission.AggregateRead;
+import static org.opentestsystem.rdw.reporting.common.model.DimensionType.Overall;
+import static org.opentestsystem.rdw.reporting.common.model.OrganizationType.State;
+import static org.opentestsystem.rdw.reporting.common.test.support.PermissionScopes.permissions;
+
+/**
+ * Yes, this test suite is anemic to say the least. However, the alt score implementation was
+ * copied from the claim score implementation so we really just need to test that the basic
+ * SQL clauses are correct. No need to redo testing of dimensions, filters, etc.
```
Comment (malaffoon) [inline aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcAltScoreReportRepositoryRST.java:33]: This.

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcClaimReportRepositoryRST.java [hunk 8]
```diff
@@ -142,7 +138,7 @@ public void itShouldReturnStateData() {
 
         final ClaimRow.Builder claimRowBuilder = ClaimRow.builder()
                 .assessment(ActiveAssessment.builder().id(-8).gradeCode("-4").subjectCode("Math").label("asmt-8").examSchoolYear(1999).build())
-                .organization(Organization.builder().name(stateName).organizationType(State).build())
+                .organization(Organization.builder().organizationType(State).build())
```
Comment (malaffoon) [inline aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcClaimReportRepositoryRST.java:141]: Did this in a few places: the state name isn't actually used nor retrieved. It was breaking things. This is left over from the tenant changes.

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcClaimReportRepositoryRST.java [hunk 9]
```diff
@@ -589,7 +585,8 @@ public void itShouldFindByQueryWithEthnicityDimension() {
                 .build(), District).getPayload();
 
         assertThat(reportRows.size()).isEqualTo(ethnicityCount * mathClaimsCount);
-        assertThat(reportRows.stream().filter(r -> (r.getMeasures().getStudentCount() > 0)).collect(toList()).size()).isEqualTo(mathClaimsCount);
+        assertThat((int) reportRows.stream().filter(r -> (r.getMeasures().getStudentCount() > 0)).count())
+            .isEqualTo(mathClaimsCount);
```
Comment (malaffoon) [inline aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcClaimReportRepositoryRST.java:589]: A common warning fix: replace `.collect().size()` with `count()`

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/sqlbuilder/impl/query/CustomAggregateReportQueryTemplateProviderRST.java [hunk 10]
```diff
@@ -35,15 +33,18 @@
...
 @Import({
-        SqlBuilderConfig.class,
-        QueryProvider.class
+    SqlBuilderConfig.class,
+    QueryProvider.class
 })
 public class CustomAggregateReportQueryTemplateProviderRST extends RepositoryBackedIT {
 
     @Autowired
-    @Qualifier("reportingSystemSettings")
```
Comment (malaffoon) [inline aggregate-service/src/test/java/org/opentestsystem/rdw/olap/sqlbuilder/impl/query/CustomAggregateReportQueryTemplateProviderRST.java:46]: This was causing the test(s) to fail, since the name of the bean was changed a while ago.

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/sqlbuilder/impl/query/CustomAggregateReportQueryTemplateProviderRST.java [hunk 11]
```diff
@@ -258,7 +256,7 @@ private static String state504Sql(final AssessmentType assessmentType) {
                 assessmentBasedLevelAndFrom(assessmentType) +
                 "JOIN asmt a ON a.id = e.asmt_id " +
                 "WHERE (( e.school_year IN (:school_years) AND a.grade_id IN (:asmt_grade_ids) AND a.subject_id in (:subject_ids) AND a.type_id = :asmt_type_id) " +
-                (assessmentType == SUMMATIVE ? "AND ( e.school_year < 2018 OR EXISTS (SELECT 1 FROM state_embargo WHERE aggregate = 0) OR 1 = :state_embargo_admin) " : "") +
+                (assessmentType == SUMMATIVE ? "AND ( e.school_year < :system_school_year OR EXISTS (SELECT 1 FROM state_embargo WHERE aggregate = 0) OR 1 = :state_embargo_admin) " : "") +
```
Comment (malaffoon) [inline aggregate-service/src/test/java/org/opentestsystem/rdw/olap/sqlbuilder/impl/query/CustomAggregateReportQueryTemplateProviderRST.java:259]: Another lingering problem that broke a while ago.

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/sqlbuilder/QueryProvider.java [hunk 12]
```diff
@@ -41,18 +46,31 @@ public QueryProvider(final SqlBuilderConfig sqlBuilderConfig) {
     public SelectQuery newQuery(final String templateName, final Iterable<String> addons) {
         checkArgument(sqlBuilderConfig.getTemplates() != null && sqlBuilderConfig.getTemplates().containsKey(templateName), "unknown sql builder template name " + templateName);
 
+        // get the specified report template, checking that it has some clauses
         final SqlTemplate template = sqlBuilderConfig.getTemplates().get(templateName);
         checkArgument(template.getClauses() != null, "missing clauses for template " + templateName);
```
Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/reporting/common/sqlbuilder/QueryProvider.java:52]: These changes merge the default addons which are now in a separate yml file, i.e. the large changes to those sql yml files above.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/web/converter/AltScoreAggregateReportCsvHandler.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    public boolean accept(final ReportContentResource resource, final MediaType requestedMediaType) {
+        return super.accept(resource, requestedMediaType)
+            && resource.getReport().getQuery().getType() == ReportQueryType.AltScore;
+    }
+
+    @Override
+    protected List<String> getMeasureHeaders(final AltScoreAggregateReportQuery query) {
+        return getMeasureHeaders(query, "");
```
Comment (malaffoon) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/web/converter/AltScoreAggregateReportCsvHandler.java:31]: Not sure if we will need different headers for alt-score vs. overall score, for now i'm just reusing them.

Comment (gregcharles) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/web/converter/AltScoreAggregateReportCsvHandler.java:31]: I'm not sure either. I figured the subclasses of AbstractAggregateReportCsvHander would come into play when exporting aggregate reports to CSV, but they don't as far as I can tell. I couldn't actually find any scenario that caused them to be called. I traced them up to org.springframework.http.converter.ObjectToStringHttpMessageConverter#writeInternal but then lost the trail.

The keys "report.aggregate.export.header..." are also strange. There are no label codes in the subject_translation table that start like this, and that's where we normally get the header text for Claim and Alt Score reports.

Comment (malaffoon) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/web/converter/AltScoreAggregateReportCsvHandler.java:31]: The non-subject labels are in the two `en.json` files, in this case the one in reporting common.
